cpp/libs/gst_nmos_sender_plugin/gst_nmos_sender_plugin.c

```diff
@@ -0,0 +1,308 @@
+
+/* GStreamer
+ *
+ * Copyright (C) <2024> Bisect Lda
+ * @author: Luís Ferreira <luis.ferreira@bisect.pt>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public
+ * License along with this library; if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Suite 500,
+ * Boston, MA 02110-1335, USA.
+ */
+
+#include <gst/gst.h>
+#include <gst/gstpad.h>
+#include <ossrf_c_nmos_api.h>
+
+GST_DEBUG_CATEGORY_STATIC(gst_nmossender_debug_category);
+#define GST_CAT_DEFAULT gst_nmossender_debug_category
+#define GST_TYPE_NMOSSENDER (gst_nmossender_get_type())
+#define GST_NMOSSENDER(obj) (G_TYPE_CHECK_INSTANCE_CAST((obj), GST_TYPE_NMOSSENDER, GstNmossender))
+#define GST_NMOSSENDER_CLASS(klass) (G_TYPE_CHECK_CLASS_CAST((klass), GST_TYPE_NMOSSENDER, GstNmossenderClass))
+#define GST_IS_NMOSSENDER(obj) (G_TYPE_CHECK_INSTANCE_TYPE((obj), GST_TYPE_NMOSSENDER))
+#define GST_IS_NMOSSENDER_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE((klass), GST_TYPE_NMOSSENDER))
+
+typedef struct _GstNmossender
+{
+    GstBin parent;
+    GstElement* payloader;
+    GstElement* udpsink;
+    gchar* source_address;
+    gchar* interface_name;
+    gchar* destination_address;
+    gchar* destination_port;
+    GstCaps* caps;
+} GstNmossender;
+
+typedef struct _GstNmossenderClass
+{
+    GstBinClass parent_class;
+} GstNmossenderClass;
+
+G_DEFINE_TYPE_WITH_CODE(GstNmossender, gst_nmossender, GST_TYPE_BIN,
+                        GST_DEBUG_CATEGORY_INIT(gst_nmossender_debug_category, "nmossender", 0, "NMOS Sender Plugin"))
+
+/* Pad template */
+static GstStaticPadTemplate sink_template =
+    GST_STATIC_PAD_TEMPLATE("sink", GST_PAD_SINK, GST_PAD_REQUEST,
+                            GST_STATIC_CAPS("video/x-raw, "
+                                            "format=(string){ I420, NV12, RGB } "));
+
+/* Set properties so element variables can change depending on them */
+static void gst_nmossender_set_property(GObject* object, guint property_id, const GValue* value, GParamSpec* pspec)
+{
+    GstNmossender* self = GST_NMOSSENDER(object);
+
+    switch(property_id)
+    {
+    case 1:
```

Comment:
Can we have constants for this, so that it is harder to cause problems if we add more properties?

This makes me think of something else: I'd like those constants to be `constexpr` but this file is `.c`. Can we make it a `.cpp` and declare that needs external linkage as `extern "C"`?
That way we wouldn't be limited to develop in C.

Reply:
I was checking this out, and I noticed that the property_id is something required to assign to any property on moment of creation. Since get and set property functions are called internally by GStreamer, it needs to have specifically those entry arguments making it so if there was to implement another identifier for the switch case we would before need to translate those int's to the created enums. Is that what I'm supposed to do?

Reply:
Yes. A static_cast<> is appropriate there.